### PR TITLE
nips reproduce

### PR DESCRIPTION
**Pull request by ahn1340**
This PR adds scripts for reproducing the result (Figure 3) in the paper Efficient and Robust Automated Machine Learning (Feurer et al., 2015).

### CODE REVIEW

**Comment by ahn1340** [comment]
Hi @herilalaina, please excuse my terribly late reply. To use CV instead of holdout, you can simply specify the resampling_strategy argument to 'cv', and specify folds in resampling_strategy_arguments, as shown in the [API](https://automl.github.io/auto-sklearn/stable/api.html#autosklearn.classification.AutoSklearnClassifier). The scripts will be merged soon, and hopefully in the future, users will be able to reproduce the experiment results easily.

**Comment by ahn1340** [comment]
@mfeurer, the PR is ready for review. Travis fails due to pep8 errors from slurm_helper.py. Should I also correct those errors?

File: scripts/2015_nips_paper/setup/get_tasks.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    # query only those with NaN as evaluation_measures.
+    # tasks_d = tasks_d.query("evaluation_measures != evaluation_measures")
+
+    # query only those with holdout as the resampling startegy.
+    tasks_d = tasks_d[(tasks_d.estimation_procedure == "33% Holdout set")]
+
+    task_ids = []
+    for did in dataset_ids:
+        task_a = list(tasks_a.query("did == {}".format(did)).tid)
+        if len(task_a) > 1:  # if there are more than one task, take the first one.
```
Comment (mfeurer) [inline scripts/2015_nips_paper/setup/get_tasks.py]: Could you please get the one with the lowest task ID instead of the first one?

File: scripts/2015_nips_paper/setup/create_commands.sh [hunk 2]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env bash
+
+dir=../log_output  # working directory
+#task_ids=$(python get_tasks.py)
+#task_ids="233 236 242 244 246 248 251 252 253 254 256 258 260 261 262 266 273 275 288 2117 2118 2119 2120 2122 2123 2350 3043 3044 75090 75092 75093 75098 75099 75100 75103 75104 75105 75106 75107 75108 75111 75112 75113 75114 75115 75116 75117 75119 75120 75121 75122 75125 75126 75129 75131 75133 75136 75137 75138 75139 75140 75142 75143 75146 75147 75148 75149 75150 75151 75152 75153 75155 75157 75159 75160 75161 75162 75163 75164 75165 75166 75168 75169 75170 75171 75172 75173 75174 75175 75176 75179 75180 75182 75183 75184 75185 75186 75188 75189 75190 75191 75192 75194 75195 75196 75197 75198 75199 75200 75201 75202 75203 75204 75205 75206 75207 75208 75209 75210 75212 75213 75216 75218 75220 75222 75224 75226 75228 75229 75233 75238 75240 75244 75245 75246 75247 75248 75249 75251 "
+#seeds="0 1 2 3 4 5 6 7 8 9 "
+#time_limit=3600
+#per_runtime_limit=360
+task_ids="233 236 242 244 246 248 251 252"
```
Comment (mfeurer) [inline scripts/2015_nips_paper/setup/create_commands.sh]: Please do not check in test values.

File: scripts/2015_nips_paper/setup/create_commands.sh [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+seeds="2 3 4 "
+time_limit=60
+per_runtime_limit=6
+
+#rm -r ../commands.txt
```
Comment (mfeurer) [inline scripts/2015_nips_paper/setup/create_commands.sh]: Are these comments necessary?

File: scripts/2015_nips_paper/setup/create_commands.sh [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+                --per-runtime-limit $per_runtime_limit "
+            elif [ "$model" == "metalearning" ]; then
+                cmd="python run_auto_sklearn.py --working-directory $dir/$model --task-id $task_id \
+                -s $seed --output-file score_$model.csv --model $model --time-limit $time_limit \
+                --per-runtime-limit $per_runtime_limit "
+                cmd+="; python run_auto_sklearn.py --working-directory $dir/$model --task-id $task_id \
+                -s $seed --output-file score_meta_ensemble.csv --model meta_ensemble --time-limit $time_limit \
+                --per-runtime-limit $per_runtime_limit "
+            fi
+            # Create commands.txt in the 2015_nips_paper folder and not in setup.
```
Comment (mfeurer) [inline scripts/2015_nips_paper/setup/create_commands.sh]: not in setup -> not in the setup folder.

File: scripts/2015_nips_paper/setup/create_commands.sh [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+# Create commands
+echo "creating commands.txt..."
+for seed in $seeds; do
+    for task_id in $task_ids; do
+        for model in vanilla metalearning; do
+            # store vanilla and ensemble in one folder and meta and meta_ens in another.
+            if [ "$model" == "vanilla" ]; then
+                cmd="python run_auto_sklearn.py --working-directory $dir/$model --task-id $task_id \
+                -s $seed --output-file score_$model.csv --model $model --time-limit $time_limit \
```
Comment (mfeurer) [inline scripts/2015_nips_paper/setup/create_commands.sh]: Could you please make the command creation consistent and use either the variable or a constant value in the csv name?

File: scripts/2015_nips_paper/run/run_commands.sh [hunk 6]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
+
+# Run all commands in commands.txt. Modify this to run on a cluster.
+cat "../commands.txt" | while read line; do eval "$line"; done
```
Comment (mfeurer) [inline scripts/2015_nips_paper/run/run_commands.sh]: This would only work sequentially because only then it is guaranteed that the ensemble creation for an ensemble of size 50 is done after fitting the model. If all commands were started in parallel, this would not work I guess. If that's the case, could you please make sure that the two different command types are written in two separate files and update the readme?

File: scripts/2015_nips_paper/run/run_without_metalearning.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
+import argparse
```
Comment (mfeurer) [inline scripts/2015_nips_paper/run/run_without_metalearning.py]: This file is almost equal to the one above. Could you please combine them into a single file to keep the amount of code to reproduce the experiments shorter?

File: scripts/2015_nips_paper/run/score_ensemble.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+                     delete_output_folder_after_terminate=True,
+                     shared_mode=True)
+    valid_labels = backend.load_targets_ensemble()
+    # validation_files[i]: validation prediction of i-th model
+    # validation_files[i][j]:
+    # validation files contain predictions of models
+    score = balanced_accuracy
+
+    for i in range(len(validation_files)):
+        loss = 1 - score(valid_labels, validation_files[i][0])
```
Comment (mfeurer) [inline scripts/2015_nips_paper/run/score_ensemble.py]: Could you please adapt this to work both with scores and losses?

Comment (mfeurer) [reply scripts/2015_nips_paper/run/score_ensemble.py]: (as well as below)

File: scripts/2015_nips_paper/plot/plot_ranks.py [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+    # return the trajectories over seeds (series object)
+    return series
+
+
+def main():
+    # name of the file where the plot is stored
+    saveto = "../plot.png"
+    # runtime of each experiment
+    max_runtime = 60
```
Comment (mfeurer) [inline scripts/2015_nips_paper/plot/plot_ranks.py]: Could you please change this to the actual value used in the experiment?

File: scripts/2015_nips_paper/plot/plot_ranks.py [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+            # current model and current task.
+            trajectory = fill_trajectory(performance_list, time_list)
+            trajectories.append(trajectory)
+
+        # list[list[pd.Series]]
+        trajectories_all_models.append(trajectories)
+
+    # Maps from task_id to model to pd.Series object
+    # containing trajectories across different seeds.
+    trajectories_by_task = {t: {m: trajectories_all_models[i][j]
```
Comment (mfeurer) [inline scripts/2015_nips_paper/plot/plot_ranks.py]: If I see it correctly you could directly create `all_trajectories` instead of creating this intermediate data structure, right?

File: scripts/2015_nips_paper/plot/plot_ranks.py [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+    for model in model_list:
+        trajectories = [
+            trajectories_by_task[task][model] for task in
+            trajectories_by_task
+        ]
+        all_trajectories.append(trajectories)
+
+    all_rankings = []
+    n_iter = 500  # number of bootstrap samples to use for estimating the ranks.
+    n_tasks = len(trajectories_by_task)
```
Comment (mfeurer) [inline scripts/2015_nips_paper/plot/plot_ranks.py]: replace by `len(task_list)`?

File: scripts/2015_nips_paper/plot/plot_ranks.py [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+    for i in range(n_iter):
+        pick = np.random.choice(all_trajectories[0][0].shape[1],
+                                size=(len(model_list, )))
```
Comment (mfeurer) [inline scripts/2015_nips_paper/plot/plot_ranks.py]: Could you please explain the comma after model_list? Shouldn't it be `size=(len(model_list), )`?

File: scripts/2015_nips_paper/plot/plot_ranks.py [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+
+                time_steps = [float(i.strip()) for i in csv_data[:, 0]]
+                assert time_steps[0] == 0
+
+                performance_list.append(data)
+                time_list.append(time_steps)
+
+            # trajectory is the pd.Series object containing all seed runs of the
+            # current model and current task.
+            trajectory = fill_trajectory(performance_list, time_list)
```
Comment (mfeurer) [inline scripts/2015_nips_paper/plot/plot_ranks.py]: Don't you have to do this once the data for all models is loaded, i.e. outside the nested for loops?

File: scripts/2015_nips_paper/setup/create_commands.sh (hunk 2 above)
Comment (ahn1340) [reply scripts/2015_nips_paper/setup/create_commands.sh]: Yes, my bad. I will remove them.

File: scripts/2015_nips_paper/run/run_without_metalearning.py (hunk 7 above)
Comment (ahn1340) [reply scripts/2015_nips_paper/run/run_without_metalearning.py]: I agree that they can be merged into one file. In fact, I would merge run_with_metalearning, run_without_metalearning, and run_auto_sklearn into one file as most of what run_auto_sklearn does is just calling the other two files.

File: scripts/2015_nips_paper/run/run_commands.sh (hunk 6 above)
Comment (mfeurer) [reply scripts/2015_nips_paper/run/run_commands.sh]: I see, could you please add a comment about this to `create_commands.sh` and `run_commands.sh`?

File: scripts/2015_nips_paper/run/score_ensemble.py (hunk 8 above)
Comment (ahn1340) [reply scripts/2015_nips_paper/run/score_ensemble.py]: Hi Matthias, do I understand correctly that you want the file to be able to deal with both the loss, which is actually balanced error rate (1 - balanced accuracy) in this case to be minimized, and score, which would be balanced accuracy to be maximized, by allowing users to specify it through the commandline argument?

File: scripts/2015_nips_paper/plot/plot_ranks.py (hunk 12 above)
Comment (ahn1340) [reply scripts/2015_nips_paper/plot/plot_ranks.py]: From the code you sent me, the comma was there and I did not change it as size=(len(model_list, )) does the same thing as size=len(model_list). I believe that size should simply be a scalar (4, specifically), and therefore it could be just len(model_list) without any comma. Would you agree?

File: scripts/2015_nips_paper/run/score_ensemble.py (hunk 8 above)
Comment (mfeurer) [reply scripts/2015_nips_paper/run/score_ensemble.py]: Hm, thinking about it again, maybe it is sufficient to point out that it only works with balanced accuracy.

File: scripts/2015_nips_paper/plot/plot_ranks.py (hunk 12 above)
Comment (mfeurer) [reply scripts/2015_nips_paper/plot/plot_ranks.py]: Yes, could you then please remove all commas from that argument?

File: scripts/2015_nips_paper/plot/plot_ranks.py (hunk 13 above)
Comment (ahn1340) [reply scripts/2015_nips_paper/plot/plot_ranks.py]: Do I understand you correctly that the code should fill all trajectories across all models to share the same time step indices? That happens in line 142-146, where fillna() is applied to the picked trajectories, and in line 156, where fillna() is applied to the computed ranks. This comes mostly from the code you sent me, which I admittedly tried not to alter too much. However, I see that it makes more sense and intuitive to do fill_trajectory once and for all models when creating the dataframe object, instead of filling the trajectories many times later. Should I modify the code in such a way?

Comment (mfeurer) [reply scripts/2015_nips_paper/plot/plot_ranks.py]: Okay, thanks for the explanation, that makes sense.

File: scripts/2015_nips_paper/run/run_commands.sh (hunk 6 above)
Comment (ahn1340) [reply scripts/2015_nips_paper/run/run_commands.sh]: Yep, I will do that

File: scripts/2015_nips_paper/plot/plot_ranks.py (hunk 10 above)
Comment (ahn1340) [reply scripts/2015_nips_paper/plot/plot_ranks.py]: Yep, I will fix that

File: scripts/2015_nips_paper/run/score_ensemble.py (hunk 8 above)
Comment (ahn1340) [reply scripts/2015_nips_paper/run/score_ensemble.py]: Okay, I will then write that on Readme.

**Comment by ahn1340** [comment]
@mfeurer, the PR is ready for review again. Could you have a look at it?
